### DISCUSSION TITLE
Enhance BusinessRuleAsync execution with CanExecuteInParallel

### DISCUSSION THREAD

**Discussion by StefanOssendorf** [MarimerLLC/csla#3856] [Ideas]
This is probably something after #3736 

Add a new property `public bool CanExecuteInParallel { get; protected set;} = true;` in `BusinessRuleAsync`.
With this flag the rule execution can be tweaked to disallow the exeuction of two async rules in parallel.
Currently, as far as I know, all async rules are executed in parallel regardles of priority or other constraints.
This would definitely slow down rule execution when used, but this is opt-in so the user should be aware of slower execution.

Use case:
At work we experienced sometimes EF core exceptions due to concurrent access of the DbContext. This happend after some performance tweaks and was tracked down to async rules which needed some info from the db.
We found two options to avoid this:
1. Span a new di scope with the help of the application context
2. Create an indirection with a class inbetween which locks the access to avoid the concurrent DbContext access

Both are so so and with the above proposed we could sequentialize rules which access a shared resource.

Opinions?

**Comment by rockfordlhotka** [comment]
This makes me think of some software I wrote around 1992 (give or take) to orchestrate the running of nightly batch processing jobs.

Some jobs had to run one at a time, others could run in parallel. The jobs that ran one at a time were prioritized - some ran before the parallel jobs, some after and they all ran in a specific order. There was even a short-circuit idea, because if certain jobs failed then no other jobs should run after that point.

**Reply by StefanOssendorf** [reply]
Okay. What does that mean? Should we create an enhancement task or discard it?

**Reply by rockfordlhotka** [reply]
I think it is a good idea. I also think we should work through the design in some depth, because if we're going to create an orchestration system, we need to make sure we understand the requirements.
